### PR TITLE
11 implement mcmc stage

### PR DESCRIPTION
**Pull request by cklyne**
Implemented MCMC stage.
Implemented estimator saving.
Minor adjustments to ENV.

### CODE REVIEW

**Review by SteffenJw — APPROVED** [review]

File: bde/ml/models.py [hunk 1]
```diff
@@ -827,17 +897,137 @@ def _prior_fitting(
...
+            _, states = jax.lax.scan(
+                f=one_step,
+                init=initial_state,
+                xs=keys,
+            )
+
+            return states
+
+        sample_keys = jax.random.split(split_key, self.n_chains)
+        pmap_states = jax.pmap(
```
Comment (EmanuelSommer) [inline bde/ml/models.py:996]: This only works as long as n_chains < available devices right?

Comment (EmanuelSommer) [reply bde/ml/models.py:996]: This is not always the setting

File: bde/ml/models.py [hunk 2]
```diff
@@ -320,70 +324,103 @@ def __init__(
...
         """
+        atts = {
+            "batch_size": self.batch_size,
+            "validation_size": self.validation_size,
+            "seed": self.seed,
+            "is_fitted_": self.is_fitted_,
+            "n_features_in_": self.n_features_in_,
+        }
         children = (
             self.model_,
```
Comment (EmanuelSommer) [inline bde/ml/models.py:350]: Not saving the posterior samples?

File: bde/ml/models.py [hunk 3]
```diff
@@ -827,17 +897,137 @@ def _prior_fitting(
...
+        def logdensity_for_batch(
+            params,
+            carry: float,
+            batch: Tuple[ArrayLike, ArrayLike],
+        ) -> Tuple[float, None]:
+            x, y = batch
+            y_pred = self.model_.apply(params, x)
+            res_loss = self.loss(y, y_pred)
+            res_loss = jnp.sum(res_loss)
+            res_prior = self.log_prior(params)
```
Comment (EmanuelSommer) [inline bde/ml/models.py:928]: should not be applied n_batches many times! We do a full batch evaluation meaning we push the whole dataset at once. So we indeed can du a running sum of the log likelihood but not over the log prior over the params which are independent of the batch/dataset!

Also the loss is only proportional to the log likelihood please use the exact likelihood here. also adjust the docstring in the loss definition saying it calculates the log likelihood. it does not - it just returns something proportional to it.

File: bde/ml/models.py [hunk 4]
```diff
@@ -901,17 +1137,42 @@ def predict(
...
-        bde.utils.utils.check_predict_input(X)
-        chex.assert_equal(self.is_fitted_, True)
-        return jax.pmap(
+        bde.utils.utils.check_predict_input(X, self.is_fitted_)
+        res = jax.lax.scan(
+            f=lambda carry, params: (0.0, self.model_.apply(params, X)),
+            init=0.0,
+            xs=self.samples_,
+        )[1]
+        return res.mean(axis=0)
```
Comment (EmanuelSommer) [inline bde/ml/models.py:1146]: this mean prediction should definitely be an option but add the option with a further argument to get predictions by sampling from the fitted Gaussians. The argument could be something like `n_samples` and should indicate how many predictions should be sampled for each parameterized Gaussian (for each posterior sample). In such a way we allow sampling from the empirical posterior predictive distribution. Otherwise we would discard any information about the learned log variance. Same goes for de and credibility intervals

File: bde/ml/models.py (hunk 3 above)
Comment (EmanuelSommer) [reply bde/ml/models.py:928]: also you apply log to the learned std - shouldn't it be the exp?

File: pyproject.toml [hunk 5]
```diff
@@ -28,43 +30,43 @@ Issues = "https://github.com/applied-ml-bde/bde/issues"
 [tool.setuptools_scm]
 version_file = "bde/_version.py"
 
-[tool.poetry]
```
Comment (EmanuelSommer) [inline pyproject.toml:31]: I have some setup difficulties with pixi. Please provide instructions in the Readme!

**Review by EmanuelSommer — CHANGES_REQUESTED** [review]
Only had time for a rather superficial look. Nevertheless pls have a look at my comments and address/clarify appropriately. I like the overall structure a lot. We just have to make sure the details especially regrading the log unnormalized posterior are correct.

**Comment by EmanuelSommer** [comment]
@SteffenJw do you call that a review? :thinking:

File: bde/ml/models.py (hunk 2 above)
Comment (cklyne) [reply bde/ml/models.py:350]: The `FullyConnectedEstimator` class is a regular neural network with no sampling. The `BDEEstinator` class implements the sampling functionality and it does save the samples when serializing (line 785).

File: pyproject.toml (hunk 5 above)
Comment (cklyne) [reply pyproject.toml:31]: I updated the [development setup](https://github.com/applied-ml-bde/bde/tree/11-implement-mcmc-stage?tab=readme-ov-file#development-setup) section in the readme file.
Do you think this is clear enoguh, or do you recommend I change it up?

Comment (EmanuelSommer) [reply pyproject.toml:31]: It's ok. @SteffenJw should be able to reproduce without problems. Is this the case?

Comment (SteffenJw) [reply pyproject.toml:31]: > It's ok. @SteffenJw should be able to reproduce without problems. Is this the case?

yes, it works. I had redo it on a different machine because bc my pc needs new hardware. 
I just used the install commands provided in the link. 
Just dont do the intuitive pip install pixi. Maybe there should be a warning . As far as I understand, that installs something completely different.

Comment (SteffenJw) [reply pyproject.toml:31]: (i installed it on windows if that matters)

File: bde/ml/models.py (hunk 1 above)
Comment (cklyne) [reply bde/ml/models.py]: This issue should be fixed for the `fit` method in 5cf7e9e2f55edc080fc7ca20f857bdb22c807b60.
This fix raised a new efficiency issue in cases where `n_chains % n_devices > 0` which was only partially addressed, but we could document it as a bug for now and address it later based on time constraints.
I currently disabled parallelization for the `predict` methods altogether since jitting my solution for the `fit` method proved non-trivial. I plan to revisit it in a future PR.

File: bde/ml/models.py (hunk 4 above)
Comment (cklyne) [reply bde/ml/models.py]: By "sampling from the fitted Gaussians", do you mean that after evaluating the `mean` and `std` for each sampled netwrok we should use them to parametrize `n_mcmc_samples` Gaussian distributions, and samples from each of them `n_samples` points (in total `n_samples` $\times$ `n_mcmc_samples` points), and use these points to evaluate the mean prediction and the confidence interval?

File: bde/ml/models.py (hunk 3 above)
Comment (cklyne) [reply bde/ml/models.py]: Does applying the data in batches break the algorithm?
If it doesn't I would rather keep it since it can reduce the resource requirements.
Maybe we could keep the default batch_size value as -1 to indicate a full batch evaluation and give the user the ability to change it, since the batching is already implemented.

Comment (cklyne) [reply bde/ml/models.py]: and does the full batch requirement apply only to the mcmc-sampling, or to the DE-init as well?

Comment (EmanuelSommer) [reply bde/ml/models.py]: Batching definitely breaks the sampling algorithm yes. The user should not be able to change this as of right now with this sampler. DE warmstarts can and should be trained in a vanilla batched way.

File: bde/ml/models.py (hunk 4 above)
Comment (EmanuelSommer) [reply bde/ml/models.py]: Yes

File: bde/ml/models.py (hunk 3 above)
Comment (EmanuelSommer) [reply bde/ml/models.py]: even if it is always is single batch this is not how code should be written. It is tempting to extend it to minibatch samplers and then it will be wrong

**Review by EmanuelSommer — APPROVED** [review]
I will now approve this PR for you to being able to merge. **BUT** I did not further review any of your changes, so it is your responsibility @cklyne and @SteffenJw to make sure the methods are correct. Especially from a theoretical standpoint, it would make a lot of sense to revisit my comments before merging. And do not forget to also provide a good documentation and usage examples such that it is a breeze to test your code ;)

File: bde/ml/models.py (hunk 3 above)
Comment (cklyne) [reply bde/ml/models.py]: Right now I hardcoded `self.batch_mcmc_ = X.shape[0]` in the `fit` method, and the user has no way to change it, so the minibatch algorithm is ran on only 1 batch (batch_size = n_data_points -> 1 iter).
If this solution works with MCMC sampling I would like to keep it, since:
 - If the sampling method will be changed in the future to one which allows for minibatching, we will have the code ready to go.
 - I don't think this should effect efficiency in a noticeable way
 
@EmanuelSommer what do you think?

Comment (EmanuelSommer) [reply bde/ml/models.py]: It's ok like this, but as soon as mini batches are used the code will be methodologically wrong due to the prior. This is unfortunate.

File: bde/ml/models.py (hunk 4 above)
Comment (cklyne) [reply bde/ml/models.py]: Ok, I implemented sampling from the predicted distributions.
Now `predict` returns the mean of these samples and `predict_with_credibility_eti` returns (median, low_cred, high_cred).
